pkg/image/imageversion.go

```diff
@@ -80,38 +82,40 @@ func (c *ConformanceImageVersion) Get(client discovery.ServerVersionInterface) (
 			return "", errors.Wrap(err, "couldn't retrieve server version")
 		}
 
-		parsedVersion, err := validateVersion(version.GitVersion)
-		if err != nil {
-			return "", err
-		}
+		return conformanceTagFromSemver(version.GitVersion)
+	}
+	return string(*c), nil
+}
 
-		segments := parsedVersion.Segments()
-		if len(segments) < 2 {
-			return "", fmt.Errorf("version %q only has %d segments, need at least 2", version.GitVersion, len(segments))
-		}
+// conformanceTagFromSemver uses the gitversion to choose the proper conformance image to use.
+// Prereleases are considered, but metadata and provider-specific info is discarded.
+func conformanceTagFromSemver(gitVersion string) (string, error) {
+	parsedVersion, err := validateVersion(gitVersion)
+	if err != nil {
+		return "", err
+	}
 
-		// Temporary logic in place to truncate auto-resolved versions while we
-		// transition to upstream. If < 1.14 return 2 segments due to lag behind
-		// releases. Otherwise return 3. Use the segments instead of .major and
-		// .minor because GKE's .minor is `10+` instead of `10`.
-		if segments[0] == 1 && segments[1] < 14 {
-			return fmt.Sprintf("v%d.%d", segments[0], segments[1]), nil
-		}
+	segments := parsedVersion.Segments()
+	if len(segments) < 2 {
+		return "", fmt.Errorf("version %q only has %d segments, need at least 2", gitVersion, len(segments))
+	}
 
-		// Not sure that this would be hit but default to adding the last
-		// segment as 0 per convention (upstream + semver).
-		if len(segments) < 3 {
-			return fmt.Sprintf("v%d.%d.%d", segments[0], segments[1], 0), nil
-		}
+	// Not sure that this would be hit but default to adding the last
+	// segment as 0 per convention (upstream + semver).
+	if len(segments) < 3 {
+		return fmt.Sprintf("v%d.%d.%d", segments[0], segments[1], 0), nil
+	}
 
-		// Upstream Kubernetes publishes the conformance images for prereleases as well; we should use them
-		// to ease testing new versions.
-		if parsedVersion.Prerelease() == "" {
-			return fmt.Sprintf("v%d.%d.%d", segments[0], segments[1], segments[2]), nil
-		}
+	// Upstream Kubernetes publishes the conformance images for prereleases as well; we should use them
+	// to ease testing new versions. Some vendors seem to put their name as prerelease instead of
+	// build metadata so handle on a case-by-case basis.
+	switch pr := parsedVersion.Prerelease(); {
+	case strings.HasPrefix(pr, "rc"),
+		strings.HasPrefix(pr, "alpha"),
+		strings.HasPrefix(pr, "beta"):
```

Comment:
The fact that we'd have an ever expanding line of provider values to ignore (eks, gke, microk8s had just numbers, etc), it is easiest to just support specifically those prereleases that we know upstream does use.

Check for tags here: https://console.cloud.google.com/gcr/images/google-containers/GLOBAL/conformance?gcrImageListsize=30&gcrImageListquery=%255B%257B_22k_22_3A_22_22_2C_22t_22_3A10_2C_22v_22_3A_22_5C_22v1.17_5C_22_22%257D%255D

Reply:
hmm that's the old repository, upstream moved from `gcr.io/google-containers/conformance` to `k8s.gcr.io/conformance` in 1.19

but yes conceptually the same -- I'm okay with this approach

<details>
<summary>list of conformance tags</summary>

```
❯ skopeo list-tags docker://k8s.gcr.io/conformance
{
    "Repository": "k8s.gcr.io/conformance",
    "Tags": [
        "v1.13.0",
        "v1.13.0-alpha.2",
        "v1.13.0-alpha.3",
        "v1.13.0-beta.0",
        "v1.13.0-beta.1",
        "v1.13.0-beta.2",
        "v1.13.0-rc.1",
        "v1.13.0-rc.2",
        "v1.13.1",
        "v1.13.1-beta.0",
        "v1.13.10",
        "v1.13.10-beta.0",
        "v1.13.11",
        "v1.13.11-beta.0",
        "v1.13.12",
        "v1.13.12-beta.0",
        "v1.13.13-beta.0",
        "v1.13.2",
        "v1.13.2-beta.0",
        "v1.13.3",
        "v1.13.3-beta.0",
        "v1.13.4",
        "v1.13.4-beta.0",
        "v1.13.5",
        "v1.13.5-beta.0",
        "v1.13.6",
        "v1.13.6-beta.0",
        "v1.13.7",
        "v1.13.7-beta.0",
        "v1.13.8",
        "v1.13.8-beta.0",
        "v1.13.9",
        "v1.13.9-beta.0",
        "v1.14.0",
        "v1.14.0-alpha.0",
        "v1.14.0-alpha.1",
        "v1.14.0-alpha.2",
        "v1.14.0-alpha.3",
        "v1.14.0-beta.0",
        "v1.14.0-beta.1",
        "v1.14.0-beta.2",
        "v1.14.0-rc.1",
        "v1.14.1",
        "v1.14.1-beta.0",
        "v1.14.10",
        "v1.14.10-beta.0",
        "v1.14.11-beta.0",
        "v1.14.2",
        "v1.14.2-beta.0",
        "v1.14.3",
        "v1.14.3-beta.0",
        "v1.14.4",
        "v1.14.4-beta.0",
        "v1.14.5",
        "v1.14.5-beta.0",
        "v1.14.6",
        "v1.14.6-beta.0",
        "v1.14.7",
        "v1.14.7-beta.0",
        "v1.14.8",
        "v1.14.8-beta.0",
        "v1.14.9",
        "v1.14.9-beta.0",
        "v1.15.0",
        "v1.15.0-alpha.0",
        "v1.15.0-alpha.1",
        "v1.15.0-alpha.2",
        "v1.15.0-alpha.3",
        "v1.15.0-beta.0",
        "v1.15.0-beta.1",
        "v1.15.0-beta.2",
        "v1.15.0-rc.1",
        "v1.15.1",
        "v1.15.1-beta.0",
        "v1.15.10",
        "v1.15.10-beta.0",
        "v1.15.11",
        "v1.15.11-beta.0",
        "v1.15.12",
        "v1.15.12-beta.0",
        "v1.15.13-beta.0",
        "v1.15.2",
        "v1.15.2-beta.0",
        "v1.15.3",
        "v1.15.3-beta.0",
        "v1.15.4",
        "v1.15.4-beta.0",
        "v1.15.5",
        "v1.15.5-beta.0",
        "v1.15.6",
        "v1.15.6-beta.0",
        "v1.15.7",
        "v1.15.7-beta.0",
        "v1.15.8",
        "v1.15.8-beta.0",
        "v1.15.9",
        "v1.15.9-beta.0",
        "v1.16.0",
        "v1.16.0-alpha.0",
        "v1.16.0-alpha.1",
        "v1.16.0-alpha.2",
        "v1.16.0-alpha.3",
        "v1.16.0-beta.0",
        "v1.16.0-beta.1",
        "v1.16.0-beta.2",
        "v1.16.0-rc.1",
        "v1.16.0-rc.2",
        "v1.16.1",
        "v1.16.1-beta.0",
        "v1.16.10",
        "v1.16.10-beta.0",
        "v1.16.11",
        "v1.16.11-rc.0",
        "v1.16.12",
        "v1.16.12-rc.0",
        "v1.16.12-rc.1",
        "v1.16.13",
        "v1.16.13-rc.0",
        "v1.16.14",
        "v1.16.14-rc.0",
        "v1.16.15",
        "v1.16.15-rc.0",
        "v1.16.16-rc.0",
        "v1.16.2",
        "v1.16.2-beta.0",
        "v1.16.3",
        "v1.16.3-beta.0",
        "v1.16.4",
        "v1.16.4-beta.0",
        "v1.16.5",
        "v1.16.5-beta.0",
        "v1.16.6",
        "v1.16.6-beta.0",
        "v1.16.7",
        "v1.16.7-beta.0",
        "v1.16.8",
        "v1.16.8-beta.0",
        "v1.16.9",
        "v1.16.9-beta.0",
        "v1.17.0",
        "v1.17.0-alpha.0",
        "v1.17.0-alpha.1",
        "v1.17.0-alpha.2",
        "v1.17.0-alpha.3",
        "v1.17.0-beta.0",
        "v1.17.0-beta.1",
        "v1.17.0-beta.2",
        "v1.17.0-rc.1",
        "v1.17.0-rc.2",
        "v1.17.1",
        "v1.17.1-beta.0",
        "v1.17.10",
        "v1.17.10-rc.0",
        "v1.17.11",
        "v1.17.11-rc.0",
        "v1.17.12",
        "v1.17.12-rc.0",
        "v1.17.13",
        "v1.17.13-rc.0",
        "v1.17.14",
        "v1.17.14-rc.0",
        "v1.17.15",
        "v1.17.15-rc.0",
        "v1.17.16",
        "v1.17.16-rc.0",
        "v1.17.17",
        "v1.17.17-rc.0",
        "v1.17.18-rc.0",
        "v1.17.2",
        "v1.17.2-beta.0",
        "v1.17.3",
        "v1.17.3-beta.0",
        "v1.17.4",
        "v1.17.4-beta.0",
        "v1.17.5",
        "v1.17.5-beta.0",
        "v1.17.6",
        "v1.17.6-beta.0",
        "v1.17.7",
        "v1.17.7-rc.0",
        "v1.17.8",
        "v1.17.8-rc.0",
        "v1.17.8-rc.1",
        "v1.17.9",
        "v1.17.9-rc.0",
        "v1.18.0",
        "v1.18.0-alpha.0",
        "v1.18.0-alpha.1",
        "v1.18.0-alpha.2",
        "v1.18.0-alpha.3",
        "v1.18.0-alpha.5",
        "v1.18.0-beta.0",
        "v1.18.0-beta.1",
        "v1.18.0-beta.2",
        "v1.18.0-rc.1",
        "v1.18.1",
        "v1.18.1-beta.0",
        "v1.18.10",
        "v1.18.10-rc.0",
        "v1.18.11",
        "v1.18.11-rc.0",
        "v1.18.12",
        "v1.18.12-rc.0",
        "v1.18.13",
        "v1.18.13-rc.0",
        "v1.18.14",
        "v1.18.14-rc.0",
        "v1.18.15",
        "v1.18.15-rc.0",
        "v1.18.16",
        "v1.18.16-rc.0",
        "v1.18.17",
        "v1.18.17-rc.0",
        "v1.18.18-rc.0",
        "v1.18.2",
        "v1.18.2-beta.0",
        "v1.18.3",
        "v1.18.3-beta.0",
        "v1.18.4",
        "v1.18.4-rc.0",
        "v1.18.5",
        "v1.18.5-rc.0",
        "v1.18.5-rc.1",
        "v1.18.6",
        "v1.18.6-rc.0",
        "v1.18.7",
        "v1.18.7-rc.0",
        "v1.18.8",
        "v1.18.8-rc.0",
        "v1.18.9",
        "v1.18.9-rc.0",
        "v1.19.0",
        "v1.19.0-alpha.0",
        "v1.19.0-alpha.1",
        "v1.19.0-alpha.2",
        "v1.19.0-alpha.3",
        "v1.19.0-beta.0",
        "v1.19.0-beta.1",
        "v1.19.0-beta.2",
        "v1.19.0-rc.0",
        "v1.19.0-rc.1",
        "v1.19.0-rc.2",
        "v1.19.0-rc.3",
        "v1.19.0-rc.4",
        "v1.19.1",
        "v1.19.1-rc.0",
        "v1.19.10-rc.0",
        "v1.19.2",
        "v1.19.2-rc.0",
        "v1.19.3",
        "v1.19.3-rc.0",
        "v1.19.4",
        "v1.19.4-rc.0",
        "v1.19.5",
        "v1.19.5-rc.0",
        "v1.19.6",
        "v1.19.6-rc.0",
        "v1.19.7",
        "v1.19.7-rc.0",
        "v1.19.8",
        "v1.19.8-rc.0",
        "v1.19.9",
        "v1.19.9-rc.0",
        "v1.20.0",
        "v1.20.0-alpha.0",
        "v1.20.0-alpha.1",
        "v1.20.0-alpha.2",
        "v1.20.0-alpha.3",
        "v1.20.0-beta.0",
        "v1.20.0-beta.1",
        "v1.20.0-beta.2",
        "v1.20.0-rc.0",
        "v1.20.1",
        "v1.20.1-rc.0",
        "v1.20.2",
        "v1.20.2-rc.0",
        "v1.20.3",
        "v1.20.3-rc.0",
        "v1.20.4",
        "v1.20.4-rc.0",
        "v1.20.5",
        "v1.20.5-rc.0",
        "v1.20.6-rc.0",
        "v1.21.0-alpha.0",
        "v1.21.0-alpha.1",
        "v1.21.0-alpha.2",
        "v1.21.0-alpha.3",
        "v1.21.0-beta.0",
        "v1.21.0-rc.0",
        "v1.22.0-alpha.0"
    ]
}
```

</summary>

Reply:
👍 I need to update my bookmarks, thanks for the link